backend/debug/dwarf/dwarf_low/debug_frame_section.ml

```diff
@@ -0,0 +1,385 @@
+      Numbers.Uint64.of_nonnegative_int_exn register |> Dwarf_value.uleb128 ]
+  | Same_value register ->
+    [ Numbers.Uint8.of_nonnegative_int_exn 0x08 |> Dwarf_value.uint8;
+      Numbers.Uint64.of_nonnegative_int_exn register |> Dwarf_value.uleb128 ]
+  | Register (register1, register2) ->
+    [ Numbers.Uint8.of_nonnegative_int_exn 0x09 |> Dwarf_value.uint8;
+      Numbers.Uint64.of_nonnegative_int_exn register1 |> Dwarf_value.uleb128;
+      Numbers.Uint64.of_nonnegative_int_exn register2 |> Dwarf_value.uleb128 ]
+  | Remember_state ->
+    [Numbers.Uint8.of_nonnegative_int_exn 0x0a |> Dwarf_value.uint8]
+  | Restore_state ->
+    [Numbers.Uint8.of_nonnegative_int_exn 0x0b |> Dwarf_value.uint8]
+  | Def_cfa (register, offset) ->
+    [ Numbers.Uint8.of_nonnegative_int_exn 0x0c |> Dwarf_value.uint8;
+      Numbers.Uint64.of_nonnegative_int_exn register |> Dwarf_value.uleb128;
+      Numbers.Uint64.of_nonnegative_int_exn offset |> Dwarf_value.uleb128 ]
+  | Def_cfa_register register ->
+    [ Numbers.Uint8.of_nonnegative_int_exn 0x0d |> Dwarf_value.uint8;
+      Numbers.Uint64.of_nonnegative_int_exn register |> Dwarf_value.uleb128 ]
+  | Def_cfa_offset offset ->
+    [ Numbers.Uint8.of_nonnegative_int_exn 0x0e |> Dwarf_value.uint8;
+      Numbers.Uint64.of_nonnegative_int_exn offset |> Dwarf_value.uleb128 ]
+  | Offset_extended_sf (register, offset) ->
+    [ Numbers.Uint8.of_nonnegative_int_exn 0x11 |> Dwarf_value.uint8;
+      Numbers.Uint64.of_nonnegative_int_exn register |> Dwarf_value.uleb128;
+      Numbers.Uint64.of_nonnegative_int_exn offset |> Dwarf_value.uleb128 ]
+  | Def_cfa_sf (register, offset) ->
+    [ Numbers.Uint8.of_nonnegative_int_exn 0x12 |> Dwarf_value.uint8;
+      Numbers.Uint64.of_nonnegative_int_exn register |> Dwarf_value.uleb128;
+      Int64.of_int offset |> Dwarf_value.sleb128 ]
+  | Def_cfa_offset_sf offset ->
+    [ Numbers.Uint8.of_nonnegative_int_exn 0x13 |> Dwarf_value.uint8;
+      Int64.of_int offset |> Dwarf_value.sleb128 ]
+  | Val_offset (register, factored_offset) ->
+    [ Numbers.Uint8.of_nonnegative_int_exn 0x14 |> Dwarf_value.uint8;
+      Numbers.Uint64.of_nonnegative_int_exn register |> Dwarf_value.uleb128;
+      Numbers.Uint64.of_nonnegative_int_exn factored_offset
+      |> Dwarf_value.uleb128 ]
+  | Val_offset_sf (register, factored_offset) ->
+    [ Numbers.Uint8.of_nonnegative_int_exn 0x15 |> Dwarf_value.uint8;
+      Numbers.Uint64.of_nonnegative_int_exn register |> Dwarf_value.uleb128;
+      Int64.of_int factored_offset |> Dwarf_value.sleb128 ]
+
+let instr_size instr =
+  let ( + ) = Dwarf_int.add in
+  List.fold_left
+    (fun size value -> size + Dwarf_value.size value)
+    (Dwarf_int.zero ()) (encode instr)
+
+let instructions_size instructions =
+  let ( + ) = Dwarf_int.add in
+  List.fold_left
+    (fun size instr -> size + instr_size instr)
+    (Dwarf_int.zero ()) instructions
+
+let emit_instructions ~asm_directives instructions =
+  List.map encode instructions
```

Comment:
If we're never going to actually store the dwarf values, would it be possible to emit them directly instead of creating a list of list?
Also, you might want to create helpers function to avoid all of the duplicated `fun x -> Numbers.Uint64.of_nonnegative_int_exn x |> Dwarf_value.uleb128;`